authz/rbac_translator.go

```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package authz exposes methods to manage authorization within gRPC.
+package authz
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	v3rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	v3routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	v3matcherpb "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+)
+
+type header struct {
+	Key    *string   `json:"key"`
+	Values *[]string `json:"values"`
+}
+
+type peer struct {
+	Principals *[]string `json:"principals"`
+}
+
+type request struct {
+	Paths   *[]string `json:"paths"`
+	Headers *[]header `json:"headers"`
+}
+
+type rule struct {
+	Name    *string  `json:"name"`
+	Source  *peer    `json:"source"`
+	Request *request `json:"request"`
+}
+
+// Represents the SDK authorization policy provided by user.
+type authorizationPolicy struct {
+	Name       *string `json:"name"`
+	DenyRules  *[]rule `json:"deny_rules"`
+	AllowRules *[]rule `json:"allow_rules"`
+}
+
+func principalAny() *v3rbacpb.Principal {
+	return &v3rbacpb.Principal{
```

Comment:
Good catch. I thot I was using it in more than one place. Updated code.

Reply:
Yeah, I think it's usually fine to directly return the same struct multiple times, even when each of them might take multiple lines.
This might also apply to other functions, like `principalOr`, `principalAnd`, `permissionOr`, etc.